server.go

```diff
@@ -151,6 +152,12 @@ func (s *Server) handleConn(c *Conn) error {
 		}
 	}
 
+	// limit connections
+	if s.MaxConnections > 0 && len(s.conns) > s.MaxConnections {
+		c.Reject()
+		return nil
+	}
```

Comment:
Thanks for the correction.
Please note that this version has a bug, though: It does never call `c.Close()` because the `defer` block is declared after the `c.Reject(); return nil`.

Reply:
`c.Reject()` already calls `c.Close()`. That's why it's before the defer block.

Reply:
I'd recommend writing a test for it, but I'm not sure about the test implementation. I see that there is a general `testServer` function is `server_test.go` but it only dials a single connection. To test the connection limit, you should dial multiple connections, assert that the first one replies with `EHLO` and the exceeded one replies with the expected code.

You should also look at how other mail servers implement this and how it plays with go's `net/smtp.NewClient()`. I'd expect `smtp.NewClient()` to return the 421 error since it reads the first line from server.

Reply:
I just checked c.Reject() an deleted my comment. :-)

Reply:
> Thanks for the correction. Please note that this version has a bug, though: It does never call `c.Close()` because the `defer` block is declared after the `c.Reject(); return nil`.

You might be right, in order to preserve the original server behavour (always a `defer c.Close()` before attempting any IO on the connection), the there should be a `defer c.Close()` somewhere before `c.Reject()` so the reject can be after the defer block (the map won't contain the connection anyway). Perhaps also `SetDeadline` should be used for the reject.

Reply:
Maybe it would work best if the built in `c.Reject()` was not used:
```go
// limit connections
if maxConnsExceeded {
	defer c.Close()
	c.writeResponse(421, EnhancedCode{4, 4, 5}, "Too busy. Try again later.")
	return nil
}
```

We are before the `defer` block and handle the rejection explicitly.

Reply:
That's definitely more explicit and better to understand.

Reply:
I think one more problem is that the reject must happen after the TLS handshake...

Reply:
I've pushed an update to my PR. I hope it address all mentioned issues.
And I'll take a look at how we can test it.

Reply:
Unfortunately your solution has a race condition. Multiple goroutines may check that the limit has not been exceeded, then proceed to exceed the limit. The read and write on `s.conns` map should happen under the same lock.

Reply:
Yes, I know. I thought I'd prefer code readability over total correctness,
but probably you're right, there's no way around it.

Reply:
pushed again.

Reply:
I think I'm done with the test. Please take a look.

Reply:
Looks fine to me, it's up to @emersion to make the decision.